cmd/datasource-token-refresh/main.go

```diff
@@ -0,0 +1,238 @@
+		"Service account JSON file. Can be left empty if default credentials have sufficient permission.")
+
+	authorizationHeader = "Authorization"
+	scope               = "https://www.googleapis.com/auth/monitoring.read"
+
+	// httpHeader* are the prefixes that is used to store the name and value of of a custom headers.
+	// check https://github.com/grafana/grafana/blob/148e1c1588e9f075b14b72eb87d5463ea5bbb253/pkg/services/datasources/models.go#L34C1-L34C1 for more info.
+	httpHeaderName  = "httpHeaderName"
+	httpHeaderValue = "httpHeaderValue"
+)
+
+func main() {
+	flag.Parse()
+
+	logger := log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
+	logger = log.With(logger, "ts", log.DefaultTimestampUTC)
+	logger = log.With(logger, "caller", log.DefaultCaller)
+
+	if *grafanaToken == "" {
+		level.Error(logger).Log("msg", "--grafana-token must be set")
+		os.Exit(1)
+	}
+	if *grafanaEndpoint == "" {
+		level.Error(logger).Log("msg", "--grafana-endpoint must be set")
+		os.Exit(1)
+	}
+	if len(*datasourceIDList) == 0 {
+		level.Error(logger).Log("msg", "--datasource-id must be set")
+		os.Exit(1)
+	}
+
+	token, err := getOAuth2Token(*jsonFilePath)
+	if err != nil {
+		level.Error(logger).Log("msg", "couldn't get token", "err", err)
+		os.Exit(1)
+	}
+
+	grafanaHeader := http.Header{
+		"Content-Type":      {"application/json"},
+		authorizationHeader: {fmt.Sprintf("Bearer %s", *grafanaToken)},
+	}
+	dsSucessfullyUpdated := []string{}
+	datasourceIDs := strings.Split(*datasourceIDList, ",")
+	for _, datasourceID := range datasourceIDs {
+		// Trim whitespaces
+		datasourceID = strings.TrimSpace(datasourceID)
+		if datasourceID == "" {
+			continue
+		}
+
+		datasourceEndpoint := fmt.Sprintf("%s/api/datasources/uid/%s", *grafanaEndpoint, datasourceID)
+		data, err := getDataSourceConfig(datasourceEndpoint, grafanaHeader)
+		if err != nil {
+			level.Error(logger).Log("msg", fmt.Sprintf("error fetching datasource config of datasource id: %s", datasourceID), "err", err)
+			continue
+		}
+
+		data, err = buildUpdateDataSourceRequest(data, token)
+		if err != nil {
+			level.Error(logger).Log("msg", fmt.Sprintf("couldn't build datasource update request for datasource id: %s", datasourceID), "err", err)
+			continue
+		}
+
+		err = sendUpdateDataSourceRequest(datasourceEndpoint, data, grafanaHeader)
+		if err != nil {
+			level.Error(logger).Log("msg", fmt.Sprintf("couldn't send update datasource request to datasource id: %s", datasourceID), "err", err)
+			continue
+		}
+		dsSucessfullyUpdated = append(dsSucessfullyUpdated, datasourceID)
+	}
+	if len(dsSucessfullyUpdated) != 0 {
+		level.Info(logger).Log("msg", fmt.Sprintf("Updated grafana datasource ids: %s", dsSucessfullyUpdated))
+	}
+}
+
+// getDataSourceConfig gets the current datasource config from the Grafana API.
+func getDataSourceConfig(DataSourceEndpoint string, grafanaHeader http.Header) (map[string]interface{}, error) {
+	req, err := http.NewRequest(http.MethodGet, DataSourceEndpoint, nil)
```

Comment:
question: is there a reason we don't use Grafana's official client? https://github.com/grafana/grafana-api-golang-client